editor/src/core/workers/parser-printer/uid-fix.ts

```diff
@@ -0,0 +1,170 @@
+      const newUID = optionalMap(getUtopiaID, newElement)
+
+      if (oldUID != null && newUID != null && oldUID !== newUID) {
+        // we have a UID mismatch
+        newToOldUidMapping[newUID] = { oldUID: oldUID, newUID: newUID, pathToModify: newPath }
+      }
+    },
+  )
+
+  const newToOldUidMappingArray = Object.values(newToOldUidMapping)
+
+  if (newToOldUidMappingArray.length !== 1) {
+    // we found no uid mismatch or we found too many mismatched UIDs and so we bail out
+    return newParsed
+  } else {
+    // we found a single UID mismatch, which means there's a very good chance that it was an update element, let's fix that up
+    let workingComponents = getComponentsFromTopLevelElements(newParsed.topLevelElements)
+
+    newToOldUidMappingArray.forEach((mapping) => {
+      workingComponents = transformJSXComponentAtElementPath(
+        workingComponents,
+        mapping.pathToModify,
+        (element) => {
+          return setUtopiaIDOnJSXElement(element, mapping.oldUID)
+        },
+      )
+    })
+
+    const fixedTopLevelElements = applyUtopiaJSXComponentsChanges(
+      newParsed.topLevelElements,
+      workingComponents,
+    )
+
+    const fixedHighlightBounds = mapArrayToDictionary(
+      Object.values(newParsed.highlightBounds),
+      (highlightBound) => {
+        const uidFix = newToOldUidMapping[highlightBound.uid]
+        if (uidFix != null) {
+          return uidFix.oldUID
+        } else {
+          return highlightBound.uid
+        }
+      },
+      (highlightBound) => {
+        const uidFix = newToOldUidMapping[highlightBound.uid]
+        if (uidFix != null) {
+          return {
+            ...highlightBound,
+            uid: uidFix.oldUID,
+          }
+        } else {
+          return highlightBound
+        }
+      },
+    )
+
+    return {
+      ...newParsed,
+      topLevelElements: fixedTopLevelElements,
+      highlightBounds: fixedHighlightBounds,
+    }
+  }
+}
+
+function zipProjectContents(
+  firstTopLevelElements: Array<TopLevelElement>,
+  secondTopLevelElements: Array<TopLevelElement>,
+  onElement: (
+    firstContents: JSXElementChild | null,
+    secondContents: JSXElementChild,
+    newTemplatePath: StaticElementPath,
+  ) => void,
+): void {
+  secondTopLevelElements.forEach((newTopLevelElement, index) => {
+    if (isUtopiaJSXComponent(newTopLevelElement)) {
+      const uid = getUtopiaID(newTopLevelElement.rootElement)
+
+      const oldTopLevelElement = firstTopLevelElements.find((tle): tle is UtopiaJSXComponent => {
+        return isUtopiaJSXComponent(tle) && getUtopiaID(tle.rootElement) === uid
+      })
+
+      if (oldTopLevelElement != null) {
+        walkElementChildren(
+          TP.emptyElementPath,
+          [oldTopLevelElement.rootElement],
+          [newTopLevelElement.rootElement],
+          onElement,
+        )
+      }
+    }
+  })
+}
+
+function walkElementChildren(
+  pathSoFar: StaticElementPath,
```

Comment:
as @seanparsons observed last week, we are kind of using the TemplatePath for two things: to point to something in the file hierarchy, and to point to something in the same file. Here I'm trying to separate those concerns and instead of using TemplatePath, I'm using a `StaticElementPath` to express that this is all local to the one file we are working with. 

Also, using a TemplatePath here would require for me to provide a fake EmptyScenePath which kind of reveals why TemplatePath is not the best tool for this job